src/helpers/await.ts

```diff
@@ -0,0 +1,62 @@
+export type Awaitable<T> = T | PromiseLike<T>;
+
+export interface Signal {
+  signal?: AbortSignal;
+}
+
+export interface Interval {
+  interval?: number;
+}
+
+export type SignalAndInterval = Signal & Interval;
+
+export interface Wait<T = unknown> {
+  wait: (opts?: SignalAndInterval) => Promise<T>;
+}
+
+export interface AsyncData<T> {
+  done: boolean;
+  data?: T;
+}
+
+export type AsyncFunction<T> = () => Awaitable<AsyncData<T>>;
+
+export function getAbortSignal(
+  signal?: AbortSignal,
+  maxTimeout: number = 60_000
+): AbortSignal {
+  let abortSignal: AbortSignal;
+  if (signal == null) {
+    const controller = new AbortController();
+    abortSignal = controller.signal;
+    setTimeout(controller.abort.bind(controller), maxTimeout);
+  } else {
+    abortSignal = signal;
+  }
+  return abortSignal;
+}
+
+export async function getAsyncPoller<T = unknown>(
```

Comment:
This is really the only interesting function. It simply wraps a polling pattern, but rather than a timeout of some sort, it exposes the AbortController API so that callers have more control over when they want to abort. Having said that, the default behavior is simply to create an abort controller that will abort after 10 seconds.